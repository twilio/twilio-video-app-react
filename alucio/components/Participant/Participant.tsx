import React from 'react';
import ParticipantInfo from '../ParticipantInfo/ParticipantInfo';
import ParticipantTracks from '../ParticipantTracks/ParticipantTracks';
import { Participant as IParticipant } from 'twilio-video';

interface ParticipantProps {
  participant: IParticipant;
  disableAudio?: boolean;
  menu?: any;
  enableScreenShare?: boolean;
  onClick: () => void;
  isSelected: boolean;
}

export default function Participant({
  participant,
  disableAudio,
  enableScreenShare,
  onClick,
  isSelected,
  menu,
}: ParticipantProps) {
  return (
    <ParticipantInfo participant={participant} menu={menu} onClick={onClick} isSelected={isSelected}>
      <ParticipantTracks participant={participant} disableAudio={disableAudio} enableScreenShare={enableScreenShare} />
    </ParticipantInfo>
  );
}
