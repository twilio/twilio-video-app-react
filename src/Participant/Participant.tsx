import React from './node_modules/react';
import ParticipantInfo from '../ParticipantInfo/ParticipantInfo';
import ParticipantTracks from '../ParticipantTracks/ParticipantTracks';
import {
  LocalParticipant,
  RemoteParticipant,
} from './node_modules/twilio-video';

interface ParticipantProps {
  participant: LocalParticipant | RemoteParticipant;
}

export default function Participant({ participant }: ParticipantProps) {
  return (
    <ParticipantInfo participant={participant}>
      <ParticipantTracks participant={participant} />
    </ParticipantInfo>
  );
}
