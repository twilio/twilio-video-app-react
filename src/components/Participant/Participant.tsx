import React from 'react';
import ParticipantInfo from '../ParticipantInfo/ParticipantInfo';
import ParticipantTracks from '../ParticipantTracks/ParticipantTracks';
import { Participant as IParticipant } from 'twilio-video';

interface ParticipantProps {
  participant: IParticipant;
  videoOnly?: boolean;
  enableScreenShare?: boolean;
  onClick?: () => void;
  isSelected?: boolean;
  isLocalParticipant?: boolean;
  hideParticipant?: boolean;
}

export default function Participant({
  participant,
  videoOnly,
  enableScreenShare,
  onClick,
  isSelected,
  isLocalParticipant,
  hideParticipant,
}: ParticipantProps) {
  return (
    <ParticipantInfo
      participant={participant}
      onClick={onClick}
      isSelected={isSelected}
      isLocalParticipant={isLocalParticipant}
      hideParticipant={hideParticipant}
    >
      <ParticipantTracks
        participant={participant}
        videoOnly={videoOnly}
        enableScreenShare={enableScreenShare}
        isLocalParticipant={isLocalParticipant}
      />
    </ParticipantInfo>
  );
}
