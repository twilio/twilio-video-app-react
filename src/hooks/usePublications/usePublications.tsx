import { useEffect, useState } from 'react';
import { LocalTrackPublication, Participant, RemoteTrackPublication } from 'twilio-video';

type TrackPublication = LocalTrackPublication | RemoteTrackPublication;

export default function usePublications(participant: Participant) {
  const [publications, setPublications] = useState<TrackPublication[]>([]);

  useEffect(() => {
    // Reset the publications when the 'participant' variable changes.
    setPublications(Array.from(participant.tracks.values()) as TrackPublication[]);

    const publicationAdded = (publication: TrackPublication) =>
      setPublications(prevPublications => [...prevPublications, publication]);
    const publicationRemoved = (publication: TrackPublication) =>
      setPublications(prevPublications => prevPublications.filter(p => p !== publication));

    participant.on('trackPublished', publicationAdded);
    participant.on('trackUnpublished', publicationRemoved);
    return () => {
      participant.off('trackPublished', publicationAdded);
      participant.off('trackUnpublished', publicationRemoved);
    };
  }, [participant]);

  return publications;
}
