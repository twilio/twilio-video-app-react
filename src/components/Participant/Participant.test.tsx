import React from 'react';
import Participant from './Participant';
import usePublications from '../../hooks/usePublications/usePublications';
import { shallow } from 'enzyme';

jest.mock('../../hooks/usePublications/usePublications', () =>
  jest.fn(() => [{ trackSid: 0 }, { trackSid: 1 }])
);

describe('the Participant component', () => {
  it('should render an array of publications', () => {
    const wrapper = shallow(
      <Participant participant={'mockParticipant' as any} />
    );
    expect(usePublications).toHaveBeenCalledWith('mockParticipant');
    expect(wrapper).toMatchSnapshot();
  });
});
