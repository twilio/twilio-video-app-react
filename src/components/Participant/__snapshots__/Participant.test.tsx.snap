// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the Participant component should render an array of publications 1`] = `
<Fragment>
  <Publication
    key="0"
    participant="mockParticipant"
    publication={
      Object {
        "trackSid": 0,
      }
    }
  />
  <Publication
    key="1"
    participant="mockParticipant"
    publication={
      Object {
        "trackSid": 1,
      }
    }
  />
</Fragment>
`;
